### ISSUE TITLE
Port fleet to AWS

### ISSUE THREAD

**Issue by jakubgs** [status-im/infra-nimbus#3]
Nimbus team has received some perks from the Ethereum Foundation. To make use of these resources I want to port the whole fleet from DigitalOcean to AWS.
Details:
```
You will be eligible for:
- $5,000 in AWS Promotional Credit valid for 2 years
- 1 year of AWS Business Support (up to $5,000).
- 80 credits for self-paced labs
```

**Comment by jakubgs** [comment]
With help from @ceripower we got the `nimbus@status.im` address as a Google Group:
https://groups.google.com/a/status.im/forum/#!forum/nimbus
As well as a BitWarden collection called Nimbus to share the credentials.

I will use this to create the AWS account with the credits.

**Comment by jakubgs** [comment]
I've created the AWS account and added it to BitWarden under Nimbus colection. I had to add Status credit card, but we shouldn't get any charges for a while.
I have also registered for the Ethereum Foundation credits, but we might wait a bit for it to activate:
![aws_application_submitted](https://user-images.githubusercontent.com/2212681/72062717-a32a3b00-32d8-11ea-8013-df1fcfc53267.png)

**Comment by jakubgs** [comment]
I finally got an email about our credits being FINALLY available after 14 days of waiting:
![aws_nimbus_credits_resp](https://user-images.githubusercontent.com/2212681/72978138-24ec8f00-3dd6-11ea-959e-703a4352c0f8.png)

**Comment by jakubgs** [comment]
AWS Console is still showing no credits...
![aws_no_credits](https://user-images.githubusercontent.com/2212681/73065770-ad396580-3ea4-11ea-8189-2cf8761adcce.png)

**Comment by jakubgs** [comment]
Looks like we finally got our credits:
![aws_nimbus_ef_credits](https://user-images.githubusercontent.com/2212681/73173185-0c95b080-4105-11ea-938c-5f8b19eda366.png)
Here's a Gist with a list of services available to us through this credit:
https://gist.github.com/jakubgs/94ba9655b8332d2ecdc365f8a913adea

**Comment by jakubgs** [comment]
Most importantly we have access to these ones:
```
Amazon Elastic Compute Cloud
Amazon Simple Storage Service
Amazon Route 53
```
So we can store Terraform state in S3 and deploy EC2 instances.

**Comment by jakubgs** [comment]
Considering most Nim developers are in European countries I'm going to pick `eu-central-1` as the most sensible AWS region for the Nimbus fleet.

**Comment by jakubgs** [comment]
Some notes:

* Had to disable Consul since we don't have a cluster in AWS
* Had to disable Tinc, since it needs Consul to run

Since our logs travel to Logstash via Tinc this will be a bit of a pain.
I might have to deploy a Consul cluster in `aws-eu-central-1a` to make this work.

And it seems like it will be necessary anyway because of #4.

**Comment by jakubgs** [comment]
After realizing I will need Consul to make Tinc VPN work I had to stop the work to first deploy a Consul cluster in the same AWS region. We'll also need metric and logs aggregation hosts.

**Comment by jakubgs** [comment]
I just stubbed my to on https://github.com/hashicorp/terraform/issues/17519 since TF modules can't be conditional. This means that I can't modify the [`infra-tf-multi-provider`](https://github.com/status-im/infra-tf-multi-provider) module to include [`infra-tf-amazon-web-services`](https://github.com/status-im/infra-tf-amazon-web-services) because then all the places that use it would need to deploy AWS hosts. That makes sense in `infra-hq`, but not in other repos like `infra-sqarm` or `infra-eth-cluster`.

Some possible options:

* Use the `host_count` argument for AWS module to not deploy hosts.
  - This will create other non-count resources though, which is just messy.
* Create the Consul hosts in `infra-hq` without using `infra-tf-multi-provider`
  - Less messy, but kinda custom, but would definitely work.
* Drop Tinc and encode Nimbus hosts in an [AWS VPC](https://aws.amazon.com/vpc/) to secure log streaming.
  - This avoids having to setup Consul, and should give us the same security.

I'm not THAT familiar with AWS VPCs but I'm leaning towards that solution because it's more self-contained compared to the other two.

**Comment by jakubgs** [comment]
I've went with the third option and configured a VPC for all hosts: 8c55671a277557eef6f464e16fe1a743ff886e4f

Now I can access the ports I've opened on the VPC Security Group, but none of the others:
```
 $ sudo nmap -Pn -p22,80,443,9000 master-01.aws-eu-central-1a.nimbus.test
Nmap scan report for ec2-18-184-88-59.eu-central-1.compute.amazonaws.com (18.184.88.59)

PORT     STATE  SERVICE
22/tcp   open   ssh
80/tcp   closed http
443/tcp  closed https
9000/tcp closed cslistener
```
And hosts within the VPC and Subnet can access each others ports without issues:
```
admin@master-01.aws-eu-central-1a.nimbus.test:~ % sudo nmap -Pn -p 22,1234 10.0.1.38
Nmap scan report for ec2-3-120-248-29.eu-central-1.compute.amazonaws.com (10.0.1.38)

PORT     STATE  SERVICE
22/tcp   open   ssh
1234/tcp closed hotline
```
So we should be able to stream logs without the need for a Tinc VPN.

**Comment by stefantalpalaru** [comment]
I think we need more ports opened there: https://github.com/status-im/infra-nimbus/blob/0306223dfe6a946425aa687557a57060fc9f298c/ansible/main.yml#L22

Testnet0 uses 9000 and 9001 while testnet1 uses 9100 and 9101.

**Comment by jakubgs** [comment]
@stefantalpalaru not sure what you mean. They should be open.

Also, I'm getting this in the logs:
```
{"lvl":"ERR","ts":"2020-01-28 21:26:23+00:00","msg":"Failed to read genesis file","topics":"beacnde","tid":9,"err":"cannot open: /root/.cache/nimbus/BeaconNode/genesis.ssz"}
```
Which makes no sense because it managed to create `privkey.protobuf` in the same folder.

**Comment by jakubgs** [comment]
I've deployed a Consul cluster in AWS. Now I can get Tinc working there:
```
 $ ansible localhost -m debug -a 'var=groups["aws-eu-central-1a"]'
localhost | SUCCESS => {
    "groups[\"aws-eu-central-1a\"]": [
        "node-01.aws-eu-central-1a.consul.hq",
        "node-02.aws-eu-central-1a.consul.hq",
        "node-03.aws-eu-central-1a.consul.hq"
    ]
}
```
```
admin@node-01.aws-eu-central-1a.consul.hq:~ % consul members
Node                                 Address              Status  Type    Build  Protocol  DC                 Segment
node-01.aws-eu-central-1a.consul.hq  35.159.37.162:8301   alive   server  1.6.2  2         aws-eu-central-1a  <all>
node-02.aws-eu-central-1a.consul.hq  18.197.107.251:8301  alive   server  1.6.2  2         aws-eu-central-1a  <all>
node-03.aws-eu-central-1a.consul.hq  3.121.226.99:8301    alive   server  1.6.2  2         aws-eu-central-1a  <all>
```

**Comment by jakubgs** [comment]
Fucking lol. I just figured out why outbound traffic stopped working on my Consul hosts.
I forgot that our Tinc VPN uses the same address space as I picked for AWS VPC: `10.0.0.0/8`
```
admin@node-01.aws-eu-central-1a.consul.hq:~ % ip route
default via 10.0.1.1 dev ens5 proto dhcp src 10.0.1.46 metric 100 
10.0.0.0/8 dev tun0 proto kernel scope link src 10.4.0.1 
10.0.1.0/24 dev ens5 proto kernel scope link src 10.0.1.46 
10.0.1.1 dev ens5 proto dhcp scope link src 10.0.1.46 metric 100 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1
```
The `10.0.0.0/8 dev tun0` fucks up all outgoing traffic.
I have to change our VPC CIDR block to be `172.16.0.0 - 172.31.255.255` (`172.16/12` prefix).

https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Subnets.html

**Comment by jakubgs** [comment]
Here's the fix, but it might cause me to have to re-create all instances: c417b90faf8cc033905c80fe1e8952337ee034dc

**Comment by jakubgs** [comment]
There we go:
```
  # module.nimbus_master.aws_instance.host[0] must be replaced
...
      ~ subnet_id                    = "subnet-002740cb9f595fbd1" -> (known after apply) # forces replacement
```
Looks like I'll have to kill and re-create all hosts.

**Comment by jakubgs** [comment]
I also found a nice bug where `aws_internet_gateway` wasn't created before the hosts so Ansible provisioning couldn't work because of lack of SSH access to the hosts. Fixed in 0ca161866cfa006bd51ce3d40570937bfa6e97a0.

**Comment by jakubgs** [comment]
The Nimbus hosts are up 
https://consul.status.im/ui/aws-eu-central-1a/nodes?filter=nimbus
AWS EC2 Console:
![aws_nimbus_test_fleet](https://user-images.githubusercontent.com/2212681/73440251-3f33e900-4351-11ea-8293-5f8a3eb7c3e4.png)

And `beacon-node` services are working too:
```
admin@node-01.aws-eu-central-1a.nimbus.test:~ % d
CONTAINER ID        NAMES                    IMAGE                                    CREATED             STATUS
7abbb7eac6c9        beacon-node-testnet1-1   statusteam/nimbus_beacon_node:testnet1   10 hours ago        Up 10 hours
5c673b90cbfb        beacon-node-testnet1-2   statusteam/nimbus_beacon_node:testnet1   10 hours ago        Up 10 hours
a2a62306bcdc        beacon-node-testnet0-2   statusteam/nimbus_beacon_node:testnet0   12 hours ago        Up 10 hours
6fbf15fe98af        beacon-node-testnet0-1   statusteam/nimbus_beacon_node:testnet0   12 hours ago        Up 10 hours
```
I'm going to close this and start working on #4.
